### PR TITLE
Add custom type parser config to @types/pg

### PR DESCRIPTION
**Pull request by hittingray**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] **Provide a URL to documentation or source code which provides context for the suggested changes:** https://github.com/brianc/node-postgres/pull/1825/files
- [x] **If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header:** No. The claimed version is 7.14, [but this feature was added in 7.10](https://github.com/brianc/node-postgres/blob/master/CHANGELOG.md#7100), without the relevant type definitions ever being added.

### CODE REVIEW

File: types/pg/index.d.ts [hunk 1]
```diff
@@ -54,6 +54,11 @@ export interface QueryConfig<I extends any[] = any[]> {
     name?: string;
     text: string;
     values?: I;
+    types?: CustomTypesConfig;
+}
+
+export interface CustomTypesConfig {
+    getTypeParser: (id: pgTypes.TypeId, format: pgTypes.TypeFormat) => any;
```
Comment (hittingray) [inline types/pg/index.d.ts:61]: This type definition is out of date for the latest version of `pg-types`, but `pg` v7.14 uses `pg-types` ^v2.1.0: https://github.com/brianc/node-postgres/blob/30f67bb246cabfc6a9d4cd026e6abedbcd0e2239/package.json

The bundled TypeScript declarations were added in `pg-types` v2.2.0: [This tag](https://github.com/brianc/node-pg-types/tree/d9d9dfb87eb50914043cbc178c301b7dd3b0c50d) and here is the relevant line: https://github.com/brianc/node-pg-types/blob/d9d9dfb87eb50914043cbc178c301b7dd3b0c50d/index.d.ts#L135

I have simply taken the definitions from that point in time and applied them here.

**Review by HoldYourWaffle — COMMENTED** [review]
Looks good!
I have never used this feature myself before, so I have some questions before I can give this PR my blessing:

1. The PR you linked implies that a custom type parser can also by assigned at the `ClientConfig`-level, but I don't see any typings for that. Is this another oversight?
2. In my opinion, using `any` (unless unavoidable) is [bad](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any). Would it be possible to avoid using one here?
I think adding a generic `T`-type to the interface would already be a good step. Would it be possible/reasonable to also propagate this type to the query result?

**Comment by hittingray** [comment]
1. Yep, this was an oversight on my part. I'll fix it up.
2. Hmm, this is a bit of a tricky one to be honest. I agree that using `any` is bad, but I had gone with it to match the definitions in `pg-types`. These definitions have been fixed to use a generic in `pg-types` v3.0 and onwards though. Would you be happy with a solution something like this instead? This way, we use the imported function signature of whatever `pg-types` version `pg` is using (though this does mean that while `pg` uses `pg-types` v2.x, it will have the any definition).

```typescript
export interface CustomTypesConfig {
    getTypeParser: typeof pgTypes.getTypeParser;
}
```

I have had a look through and I don't believe it is possible to propagate the results of the custom parser to the query result (this is actually what led me down this track in the first place, I was getting mistyped results).

**Comment by HoldYourWaffle** [comment]
Yeah that makes a lot of sense to me!

**Review by HoldYourWaffle — APPROVED** [review]

**Comment by hittingray** [comment]
Thanks!

Ready to merge

**Comment by hittingray** [comment]
Ready to merge

**Pull request merged**